pkg/kubeconfig/create_kubecfg.go

```diff
@@ -35,42 +35,20 @@ const DefaultKubecfgAdminLifetime = 18 * time.Hour
 func BuildKubecfg(cluster *kops.Cluster, keyStore fi.Keystore, secretStore fi.SecretStore, cloud fi.Cloud, admin time.Duration, configUser string, internal bool, kopsStateStore string, useKopsAuthenticationPlugin bool) (*KubeconfigBuilder, error) {
 	clusterName := cluster.ObjectMeta.Name
 
-	var master string
+	var server string
 	if internal {
-		master = cluster.APIInternalName()
+		server = "https://" + cluster.APIInternalName()
 	} else {
-		master = cluster.Spec.API.PublicName
-		if master == "" {
-			master = "api." + clusterName
+		if cluster.Spec.API.PublicName != "" {
+			server = "https://" + cluster.Spec.API.PublicName
+		} else {
+			server = "https://api." + clusterName
 		}
 	}
 
-	server := "https://" + master
-
-	// We use the LoadBalancer where we know the master DNS name is otherwise unreachable
-	useELBName := false
-
-	// If the master DNS is a gossip DNS name; there's no way that name can resolve outside the cluster
-	if cluster.IsGossip() {
-		useELBName = true
-	}
-
-	// If the cluster has DNS disabled, must use the load balancer name
-	if cluster.UsesNoneDNS() {
-		useELBName = true
-	}
-
-	// If the DNS is set up as a private HostedZone, but here we have to be
-	// careful that we aren't accessing the API over DirectConnect (or a VPN).
-	// We differentiate using the heuristic that if we have an internal ELB
-	// we are likely connected directly to the VPC.
-	privateDNS := cluster.Spec.Networking.Topology != nil && cluster.Spec.Networking.Topology.DNS == kops.DNSTypePrivate
-	internalELB := cluster.Spec.API.LoadBalancer != nil && cluster.Spec.API.LoadBalancer.Type == kops.LoadBalancerTypeInternal
-	if privateDNS && !internalELB {
-		useELBName = true
-	}
-
-	if useELBName {
+	// If a load balancer exists we use it, except for when an SSL certificate is set.
```

Comment:
Perhaps also use the LB when a certificate is set? The later code will cause the kubeconfig to use the secondary port when an admin cert is being written.

Reply:
Cert means that the operator set `API.PublicName` is used for it so not sure how using the LB address in kubeconfig would work.

Reply:
`API.PublicName` gets included in the SAN of the kube-apiserver cert.

Reply:
When the cert is set, there will be the `API.PublicName` pointing to the LB with cert and address, so LB itself will be used, but not via the LB address.

Reply:
`API.PublicName` will point to the LB, but the kubeconfig will be written to use port 8443. That port won't have a certificate; it will pass through to use the cert on kube-apiserver.

Reply:
Only if `admin != 0`. In any case, this is a more advanced case.

Reply:
I had thought this PR changed the path through that port remapping logic, but on a closer look it doesn't. It just makes it not use the load balancer address in some cases where it previously would. But the apiserver cert would have `API.PublicName` in a SAN so still work.

Reply:
...but this might exacerbate #14726 since that's likely an `admin != 0` case.

Reply:
Why exacerbate?
The `SSLCertificate` is an AWS only feature, so unrelated to #14726.
We always add the LB address(es) as SANs, so there's no issue when using the kOps generated cert.

Reply:
The issue could be during validation after cluster creation, when the public name is negatively cached but the load balancer name would work.

Reply:
That's why I don't want to use public names wherever possible. LB names or IP addresses are always available.
For SSLCertificate users, it's a problem that we can't solve.